### DISCUSSION TITLE
Run on parallel in a local HPC

### DISCUSSION THREAD

**Discussion by eduardogrodriguez89** [idaholab/moose#27119] [Q&A General]
When I ran a Moose parallel computation on a HPC cluster, using the following .sh file

```
#!/bin/bash
#SBATCH --job-name=erodriguez89
#SBATCH --partition=short
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=64
#SBATCH --cpus-per-task=1
#SBATCH --time=00-01:00
conda activate moose
cd ~/projects/homogenization
srun ./homogenization-opt -i ~/projects/examples/TEMP_RVE_2D.i > pantalla_c1
```

I get the following out
```

Parallelism:
Num Processors: 1
Num Threads: 1
```
I have tried various combinations of ntasks-per-node and cpus-per-task values but it always seems to run with only one. 

[https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html ](URL)  I installed moose in this way

[https://mooseframework.inl.gov/getting_started/hpc_new_users.html ](URL) I created my app in this way 

Do I have to configure something else to run?

Thanks

**Comment by milljm** [comment]
This part of your script is a bit odd:

```pre
#SBATCH --cpus-per-task=1
#SBATCH --time=00-01:00
conda activate moose                    <------------
cd ~/projects/homogenization
srun ./homogenization-opt -i ~/projects/examples/TEMP_RVE_2D.i > pantalla_c1
```
If this Conda environment contains what I think it contains, then you are attempting to use an MPI implentation supplied by our Conda packages, instead of your HPC Cluster. SLURM in this case, is limiting your process to serial operations.

Also, If you modified the above to the following instead (force using multiple cores):
```pre
#SBATCH --cpus-per-task=1
#SBATCH --time=00-01:00
conda activate moose
cd ~/projects/homogenization
mpiexec -n 64 ./homogenization-opt -i ~/projects/examples/TEMP_RVE_2D.i > pantalla_c1
```
I suspect SLRUM will complain (because you are now attempting to use an unsupported MPI implementation).


The solution here is to _not use_ Conda. Instead using your clusters specific MPI implementation. Usually in the form of `module load`.

Edit:
This will mean basically starting over unfortunately. You will need to build PETSc, libMesh, WASP, MOOSE, and your application over again (once you figure out what modules to load).

**Reply by schakrabortygithub** [reply]
I had similar experience while using `srun` to launch an MPI job by` srun -n $SLURM_NTASKS`. It does not launch the job with multiple processors. I used `mpirun` instead.
Check which wrappers are available in the `bin` of the condo `environment` that you are using for moose. E.g., at `~/miniforge/envs/moose/bin/`
If `mpirun` or `mpiexec` is there, try using it. For me, this one works without any issue:
```
#!/bin/bash
#SBATCH --time=16:00:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=64
#SBATCH --job-name moose_app # job name
#
# load modulefiles
source ~/miniforge/etc/profile.d/conda.sh
conda activate moose

# run the app
mpirun -n $SLURM_NTASKS  ~/PathToTheApp/moose_app-opt -i ~/PathToTheInputScript/InputScript.i 
```

**Reply by milljm** [reply]
While using our Conda packages is a possible solution, and sometimes works, you may miss out on the performance offered by using your HPC Cluster's.

Some HPC Clusters are simply configured to not allow any MPI wrapper other than the one purpose-built for that cluster. Ref: https://github.com/idaholab/moose/discussions/26029

**Reply by schakrabortygithub** [reply]
@milljm, that makes sense. I might be missing some efficiency from the HPC as well by not using the system-provided MPI library. I will check.
